healthx/handler_test.go

```diff
@@ -47,9 +50,30 @@ func TestHealth(t *testing.T) {
 			},
 		},
 	}
+
+	// middleware to assert and test before the request is completed
+	testMiddleware := func(t *testing.T, assertFunc func(*testing.T, http.ResponseWriter, *http.Request)) func(next http.Handler) http.Handler {
+		return func(h http.Handler) http.Handler {
+			return http.HandlerFunc(func(writer http.ResponseWriter, req *http.Request) {
+				writer.Header().Add(mockHeaderKey, mockHeaderValue)
+				assertFunc(t, writer, req)
+				h.ServeHTTP(writer, req)
+			})
+		}
+	}
+
 	router := httprouter.New()
-	handler.SetHealthRoutes(router, true)
-	handler.SetVersionRoutes(router)
+
+	handler.SetHealthRoutes(router, true,
+		WithMiddleware(testMiddleware(t, func(t *testing.T, rw http.ResponseWriter, r *http.Request) {
+			assert.Equal(t, "GET", r.Method)
+		})),
+	)
+
+	handler.SetVersionRoutes(router, WithMiddleware(testMiddleware(t, func(t *testing.T, rw http.ResponseWriter, r *http.Request) {
+		assert.Equal(t, "GET", r.Method)
+	})))
```

Comment:
Could we maybe also add a test to make sure these handlers will still work without a middleware?

Reply:
We can separate them using a sub test 

```go
t.Run("case=with middleware", func(t *testing.T){
  ...test
})
```

Reply:
want all the tests to be inside the `t.Run` ?
It'd be better with just them being in 2 different TestHealth functions?
one for with and one for without, instead of both being the same TestHealth function

Reply:
Here's a patch before I commit, doesn't make sense to keep adding commits if it's not up to mark

```diff
diff --git a/healthx/handler_test.go b/healthx/handler_test.go
index 2d4fe82..461b1a0 100644
--- a/healthx/handler_test.go
+++ b/healthx/handler_test.go
@@ -40,73 +40,125 @@ const mockHeaderKey = "middleware-header"
 const mockHeaderValue = "test-header-value"
 
 func TestHealth(t *testing.T) {
-	alive := errors.New("not alive")
-	handler := &Handler{
-		H:             herodot.NewJSONWriter(nil),
-		VersionString: "test version",
-		ReadyChecks: map[string]ReadyChecker{
-			"test": func(r *http.Request) error {
-				return alive
+	t.Run("case:without middleware", func(t *testing.T) {
+		// separate set of `handler` and `alive` as the `alive` would need to be reset after the last test in the above
+		alive := errors.New("not alive")
+		handler := &Handler{
+			H:             herodot.NewJSONWriter(nil),
+			VersionString: "test version",
+			ReadyChecks: map[string]ReadyChecker{
+				"test": func(r *http.Request) error {
+					return alive
+				},
 			},
-		},
-	}
-
-	// middleware to assert and test before the request is completed
-	testMiddleware := func(t *testing.T, assertFunc func(*testing.T, http.ResponseWriter, *http.Request)) func(next http.Handler) http.Handler {
-		return func(h http.Handler) http.Handler {
-			return http.HandlerFunc(func(writer http.ResponseWriter, req *http.Request) {
-				writer.Header().Add(mockHeaderKey, mockHeaderValue)
-				assertFunc(t, writer, req)
-				h.ServeHTTP(writer, req)
-			})
 		}
-	}
 
-	router := httprouter.New()
+		router := httprouter.New()
 
-	handler.SetHealthRoutes(router, true,
-		WithMiddleware(testMiddleware(t, func(t *testing.T, rw http.ResponseWriter, r *http.Request) {
+		handler.SetHealthRoutes(router, true)
+		handler.SetVersionRoutes(router)
+
+		ts := httptest.NewServer(router)
+
+		c := http.DefaultClient
+
+		var healthBody swaggerHealthStatus
+		response, err := c.Get(ts.URL + AliveCheckPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusOK, response.StatusCode)
+		require.NoError(t, json.NewDecoder(response.Body).Decode(&healthBody))
+		assert.EqualValues(t, "ok", healthBody.Status)
+
+		var versionBody swaggerVersion
+		response, err = c.Get(ts.URL + VersionPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusOK, response.StatusCode)
+		require.NoError(t, json.NewDecoder(response.Body).Decode(&versionBody))
+		require.EqualValues(t, versionBody.Version, handler.VersionString)
+
+		response, err = c.Get(ts.URL + ReadyCheckPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusServiceUnavailable, response.StatusCode)
+		out, err := ioutil.ReadAll(response.Body)
+		require.NoError(t, err)
+		assert.EqualValues(t, "ok", healthBody.Status)
+		assert.Equal(t, `{"errors":{"test":"not alive"}}`, strings.TrimSpace(string(out)))
+
+		alive = nil
+		response, err = c.Get(ts.URL + ReadyCheckPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusOK, response.StatusCode)
+		require.NoError(t, json.NewDecoder(response.Body).Decode(&versionBody))
+		require.EqualValues(t, versionBody.Version, handler.VersionString)
+
+	})
+
+	t.Run("case:with middleware", func(t *testing.T) {
+		// separate set of `handler` and `alive` as the `alive` would need to be reset after the last test in the above
+		alive := errors.New("not alive")
+		handler := &Handler{
+			H:             herodot.NewJSONWriter(nil),
+			VersionString: "test version",
+			ReadyChecks: map[string]ReadyChecker{
+				"test": func(r *http.Request) error {
+					return alive
+				},
+			},
+		}
+
+		// middleware to assert and test before the request is completed
+		testMiddleware := func(t *testing.T, assertFunc func(*testing.T, http.ResponseWriter, *http.Request)) func(next http.Handler) http.Handler {
+			return func(h http.Handler) http.Handler {
+				return http.HandlerFunc(func(writer http.ResponseWriter, req *http.Request) {
+					writer.Header().Add(mockHeaderKey, mockHeaderValue)
+					assertFunc(t, writer, req)
+					h.ServeHTTP(writer, req)
+				})
+			}
+		}
+
+		routerWithMiddleware := httprouter.New()
+
+		ts := httptest.NewServer(routerWithMiddleware)
+		c := http.DefaultClient
+
+		handler.SetHealthRoutes(routerWithMiddleware, true,
+			WithMiddleware(testMiddleware(t, func(t *testing.T, rw http.ResponseWriter, r *http.Request) {
+				assert.Equal(t, "GET", r.Method)
+			})),
+		)
+
+		handler.SetVersionRoutes(routerWithMiddleware, WithMiddleware(testMiddleware(t, func(t *testing.T, rw http.ResponseWriter, r *http.Request) {
 			assert.Equal(t, "GET", r.Method)
-		})),
-	)
-
-	handler.SetVersionRoutes(router, WithMiddleware(testMiddleware(t, func(t *testing.T, rw http.ResponseWriter, r *http.Request) {
-		assert.Equal(t, "GET", r.Method)
-	})))
-
-	ts := httptest.NewServer(router)
-	c := http.DefaultClient
-
-	var healthBody swaggerHealthStatus
-	response, err := c.Get(ts.URL + AliveCheckPath)
-	require.NoError(t, err)
-	require.EqualValues(t, http.StatusOK, response.StatusCode)
-	require.NoError(t, json.NewDecoder(response.Body).Decode(&healthBody))
-	assert.EqualValues(t, "ok", healthBody.Status)
-	assert.EqualValues(t, mockHeaderValue, response.Header.Get(mockHeaderKey))
-
-	var versionBody swaggerVersion
-	response, err = c.Get(ts.URL + VersionPath)
-	require.NoError(t, err)
-	require.EqualValues(t, http.StatusOK, response.StatusCode)
-	require.NoError(t, json.NewDecoder(response.Body).Decode(&versionBody))
-	require.EqualValues(t, versionBody.Version, handler.VersionString)
-	assert.EqualValues(t, mockHeaderValue, response.Header.Get(mockHeaderKey))
-
-	response, err = c.Get(ts.URL + ReadyCheckPath)
-	require.NoError(t, err)
-	require.EqualValues(t, http.StatusServiceUnavailable, response.StatusCode)
-	out, err := ioutil.ReadAll(response.Body)
-	require.NoError(t, err)
-	assert.EqualValues(t, "ok", healthBody.Status)
-	assert.Equal(t, `{"errors":{"test":"not alive"}}`, strings.TrimSpace(string(out)))
-	assert.EqualValues(t, mockHeaderValue, response.Header.Get(mockHeaderKey))
-
-	alive = nil
-	response, err = c.Get(ts.URL + ReadyCheckPath)
-	require.NoError(t, err)
-	require.EqualValues(t, http.StatusOK, response.StatusCode)
-	require.NoError(t, json.NewDecoder(response.Body).Decode(&versionBody))
-	require.EqualValues(t, versionBody.Version, handler.VersionString)
-	assert.EqualValues(t, mockHeaderValue, response.Header.Get(mockHeaderKey))
+		})))
+
+		var healthBody swaggerHealthStatus
+		response, err := c.Get(ts.URL + AliveCheckPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusOK, response.StatusCode)
+		require.NoError(t, json.NewDecoder(response.Body).Decode(&healthBody))
+		assert.EqualValues(t, "ok", healthBody.Status)
+
+		var versionBody swaggerVersion
+		response, err = c.Get(ts.URL + VersionPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusOK, response.StatusCode)
+		require.NoError(t, json.NewDecoder(response.Body).Decode(&versionBody))
+		require.EqualValues(t, versionBody.Version, handler.VersionString)
+
+		response, err = c.Get(ts.URL + ReadyCheckPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusServiceUnavailable, response.StatusCode)
+		out, err := ioutil.ReadAll(response.Body)
+		require.NoError(t, err)
+		assert.EqualValues(t, "ok", healthBody.Status)
+		assert.Equal(t, `{"errors":{"test":"not alive"}}`, strings.TrimSpace(string(out)))
+
+		alive = nil
+		response, err = c.Get(ts.URL + ReadyCheckPath)
+		require.NoError(t, err)
+		require.EqualValues(t, http.StatusOK, response.StatusCode)
+		require.NoError(t, json.NewDecoder(response.Body).Decode(&versionBody))
+		require.EqualValues(t, versionBody.Version, handler.VersionString)
+	})
 }

```